crates/typst/src/math/attach.rs

```diff
@@ -261,145 +261,246 @@ fn layout_attachments(
+    let t_x = pre_width - t_pre_width;
+    let b_x = pre_width - b_pre_width;
+
+    // Create the final frame.
+    let mut frame = Frame::soft(Size::new(width, height));
+    frame.set_baseline(ascent);
+    frame.push_frame(Point::new(base_x, base_y), base.into_frame());
 
-    if let Some(tr) = tr {
-        let pos = Point::new(
-            sup_delta + pre_width_max + base_width,
-            ascent - shift_up - tr.ascent(),
-        );
-        frame.push_frame(pos, tr.into_frame());
+    macro_rules! layout {
+        ($e: ident, $x: ident, $y: ident) => {
+            if let Some($e) = $e {
+                frame.push_frame(Point::new($x, $y(&$e)), $e.into_frame());
+            }
+        };
     }
 
-    if let Some(br) = br {
-        let pos = Point::new(
-            sub_delta + pre_width_max + base_width,
-            ascent + shift_down - br.ascent(),
-        );
-        frame.push_frame(pos, br.into_frame());
-    }
+    layout!(tl, tl_x, tx_y); // pre-superscript
+    layout!(bl, bl_x, bx_y); // pre-subscript
+    layout!(tr, tr_x, tx_y); // post-superscript
+    layout!(br, br_x, bx_y); // post-subscript
+    layout!(t, t_x, t_y); // upper-limit
+    layout!(b, b_x, b_y); // lower-limit
 
+    // Done! Note that we retain the class of the base.
     ctx.push(FrameFragment::new(ctx, styles, frame).with_class(base_class));
 
     Ok(())
 }
 
-fn attach_top_and_bottom(
+/// Calculate the distance each post-script extends to the right of the base's
+/// width, as well as its kerning value. Requires the distance from the base's
+/// baseline to each post-script's baseline to obtain the correct kerning value.
+/// Returns 2 tuples of two lengths, each first containing the distance the
+/// post-script extends left of the base's width and second containing the
+/// post-script's kerning value. The first tuple is for the post-superscript,
+/// and the second is for the post-subscript.
+fn compute_post_script_widths(
     ctx: &MathContext,
-    styles: StyleChain,
-    base: MathFragment,
-    t: Option<MathFragment>,
-    b: Option<MathFragment>,
-) -> (Frame, Abs) {
-    let upper_gap_min = scaled!(ctx, styles, upper_limit_gap_min);
-    let upper_rise_min = scaled!(ctx, styles, upper_limit_baseline_rise_min);
-    let lower_gap_min = scaled!(ctx, styles, lower_limit_gap_min);
-    let lower_drop_min = scaled!(ctx, styles, lower_limit_baseline_drop_min);
-
-    let mut base_offset = Abs::zero();
-    let mut width = base.width();
-    let mut height = base.height();
-
-    if let Some(t) = &t {
-        let top_gap = upper_gap_min.max(upper_rise_min - t.descent());
-        width.set_max(t.width());
-        height += t.height() + top_gap;
-        base_offset = top_gap + t.height();
-    }
+    base: &MathFragment,
+    [tr, br]: [&Option<MathFragment>; 2],
+    (tr_shift, br_shift): (Abs, Abs),
+    space_after_post_script: Abs,
+) -> ((Abs, Abs), (Abs, Abs)) {
+    let tr_values = if let Some(tr) = tr {
```

Comment:
Potentially (not sure about this suggestion), the fairly frequent `if let Some(x) = x { /* Do stuff */ } else { /* Something amount of Abs::zero()s /* }` could be deduplicated with a helper function (that effectively does what `Option::map_or_default` would do if it existed).